### PR TITLE
Georgefeatures Merge

### PR DESCRIPTION
**Pull request by kingborehaha**
Project: Add UI notifier for recently completed actions (like saving). Permit & warn project folder being same as game folder. Fixed save hotkey. Minor UI tweaks.
MSB: Fixed MSB refs for non ints. MSB refs scans through arrays and mention ref type. Fixed and enabled Navmesh rendering (for supported games). 
Other: Fixed FMG menu search not checking IDs. Added Goto row ID in Param menu.

### CODE REVIEW

File: StudioCore/Editor/TaskManager.cs [hunk 1]
```diff
@@ -12,6 +12,25 @@ public class TaskManager
         public static volatile ConcurrentDictionary<string, string> warningList = new ConcurrentDictionary<string, string>();
         private static volatile ConcurrentDictionary<string, (bool, Task)> _liveTasks = new ConcurrentDictionary<string, (bool, Task)>();
         private static int _anonIndex = 0;
+        public static string lastCompletedActionString = "";
```
Comment (katalash) [inline StudioCore/Editor/TaskManager.cs:15]: This whole block is adding UI timing specific code to the task manager, which logically should not have anything to do with how the UI is displayed.

File: StudioCore/MapStudioNew.cs [hunk 2]
```diff
@@ -1010,6 +1048,12 @@ private void Update(float deltaseconds)
             }
             ImGui.End();
 
+
+            //Global shortcut keys
+            if (InputTracker.GetControlShortcut(Key.S) && !_msbEditor.Viewport.ViewportSelected)
```
Comment (katalash) [inline StudioCore/MapStudioNew.cs:1053]: What specifically does this fix?

File: StudioCore/MapStudioNew.cs [hunk 3]
```diff
@@ -637,28 +662,30 @@ private void Update(float deltaseconds)
...
+                    {
+                        focusType = "Params";
+                    }
+                    else if (_textEditorFocused)
+                    {
+                        focusType = "Text";
                     }
-                    if (ImGui.MenuItem("Save All", "CTRL+SHIFT+S") || ((InputTracker.GetKey(Key.ShiftLeft) || InputTracker.GetKey(Key.ShiftRight)) && InputTracker.GetControlShortcut(Key.S)))
+
+                    if (ImGui.MenuItem($"Save {focusType}", "Ctrl+S") || InputTracker.GetControlShortcut(Key.S))
```
Comment (katalash) [inline StudioCore/MapStudioNew.cs:684]: I think a general "save active editor" would be better, since we will eventually go to a model where editors are instantiated.

File: StudioCore/MsbEditor/Entity.cs [hunk 4]
```diff
@@ -451,6 +451,20 @@ public void BuildReferenceMap()
                     {
                         if (p.PropertyType.IsArray)
                         {
+                            Array array = (Array)p.GetValue(WrappedObject);
+                            foreach (var i in array)
+                            {
+                                var sref = (string)i;
+                                if (sref != null && sref != "")
+                                {
+                                    var obj = Container.GetObjectByName(sref);
```
Comment (katalash) [inline StudioCore/MsbEditor/Entity.cs:460]: This is effectively O(n^2), so a more scalable linear solution might be better (guess we might need to maintain a hashmap of names or something).

File: StudioCore/MsbEditor/ModelEditorScreen.cs [hunk 5]
```diff
@@ -293,12 +293,12 @@ public void ReloadAssetBrowser()
 
         public override void Save()
         {
-            
+            TaskManager.UpdateLastActionString("Model Saving is not implemented");
```
Comment (katalash) [inline StudioCore/MsbEditor/ModelEditorScreen.cs:296]: Again this is kinda introducing more global hardcoded state, so I'd prefer to hold off on this for now. It might be good to have an actual error console down the line.

File: StudioCore/MsbEditor/PropertyEditor.cs [hunk 6]
```diff
@@ -391,7 +391,7 @@ private bool ParamRefRow(PropertyInfo propinfo, object oldval, ref object newObj
                 refs.Add(att.ParamName);
                 Editor.EditorDecorations.ParamRefText(refs);
                 ImGui.NextColumn();
-                var id = (int)oldval;
+                var id = oldval; //oldval cannot always be casted to int
```
Comment (katalash) [inline StudioCore/MsbEditor/PropertyEditor.cs:394]: this feels like a band-aid to the problem. We should ideally either convert every type, or filter out those that don't match

File: StudioCore/MsbEditor/Universe.cs [hunk 7]
```diff
@@ -752,15 +752,15 @@ public async void LoadMapAsync(string mapid)
             task = job.StartJobAsync();
             tasks.Add(task);
 
-            if (FeatureFlags.LoadDS3Navmeshes)
+            if (FeatureFlags.LoadNavmeshes)
```
Comment (katalash) [inline StudioCore/MsbEditor/Universe.cs:755]: You don't need 3 feature flags here. Just loadDS3Navmeshes and LoadNavmeshes. I'd go on to rename them LoadLegacyNavmehses and LoadHavokNavmeshes as that's what they differentiate.

File: StudioCore/ParamEditor/ParamBank.cs [hunk 8]
```diff
@@ -1018,6 +1018,7 @@ public static void SaveParams(bool loose = false, bool partialParams = false)
             {
                 SaveParamsER(partialParams);
             }
+            TaskManager.UpdateLastActionString("Saved Parameters");
```
Comment (katalash) [inline StudioCore/ParamEditor/ParamBank.cs:1021]: Again we need a better system than just some spread out state and global variable. I'd hold off on this.

File: StudioCore/ParamEditor/ParamEditorScreen.cs [hunk 9]
```diff
@@ -1049,35 +1062,63 @@ public void ParamView(bool doFocus)
                 }
                 scrollTo = 0;
 
-                List<int> pinnedRowList = new List<int>(_paramEditor._projectSettings.PinnedRows.GetValueOrDefault(activeParam, new List<int>()));
-                foreach (int rowID in pinnedRowList)
+                ImGui.Text("id VALUE | name ROW | prop FIELD VALUE | propref FIELD ROW\n | original | modified");
+                UIHints.AddImGuiHintButton("MassEditHint", ref UIHints.SearchBarHint);
+
+                //Goto ID
```
Comment (katalash) [inline StudioCore/ParamEditor/ParamEditorScreen.cs:1068]: space after "//"

File: StudioCore/ParamEditor/ParamEditorScreen.cs [hunk 10]
```diff
@@ -1049,35 +1062,63 @@ public void ParamView(bool doFocus)
...
+                        _gotoParamRow = gotorow;
+                        ImGui.CloseCurrentPopup();
+                    }
+                    ImGui.EndPopup();
                 }
 
-                ImGui.Text("id VALUE | name ROW | prop FIELD VALUE | propref FIELD ROW\n | original | modified");
-                UIHints.AddImGuiHintButton("MassEditHint", ref UIHints.SearchBarHint);
 
+                //Row ID/name search
```
Comment (katalash) [inline StudioCore/ParamEditor/ParamEditorScreen.cs:1087]: Space after "//"s

File: StudioCore/ParamEditor/ParamEditorScreen.cs [hunk 11]
```diff
@@ -1049,35 +1062,63 @@ public void ParamView(bool doFocus)
...
                 if (InputTracker.GetControlShortcut(Key.F))
                     ImGui.SetKeyboardFocusHere();
+                //ImGui.SetNextItemWidth(ImGui.GetColumnWidth(1) - 256);
```
Comment (katalash) [inline StudioCore/ParamEditor/ParamEditorScreen.cs:1090]: We should probably not introduce new commented out lines like these in the codebase, unless you have a comment explaining what it was for and why you still want to keep it around.

File: StudioCore/ParamEditor/ParamEditorScreen.cs [hunk 12]
```diff
@@ -1049,35 +1062,63 @@ public void ParamView(bool doFocus)
...
 
                 if (ImGui.IsItemActive())
                     _paramEditor._isSearchBarActive = true;
                 else
                     _paramEditor._isSearchBarActive = false;
 
+
+                ImGui.BeginChild("pinnedRows");
+
+
```
Comment (katalash) [inline StudioCore/ParamEditor/ParamEditorScreen.cs:1101]: Formatting plz. Should delete the unnecessary whitespace

File: StudioCore/ParamEditor/ParamEditorScreen.cs [hunk 13]
```diff
@@ -1111,7 +1153,20 @@ private void RowColumnEntry(string activeParam, List<PARAM.Row> p,  PARAM.Row r,
                 ImGui.PushStyleColor(ImGuiCol.Text, DIRTYCOLOUR);
             else
                 ImGui.PushStyleColor(ImGuiCol.Text, CLEANCOLOUR);
-            if (ImGui.Selectable($@"{r.ID} {Utils.ImGuiEscape(r.Name, "")}", _selection.getSelectedRows().Contains(r)))
+
+            var selected = _selection.getSelectedRows().Contains(r);
+            if (_gotoParamRow != -1)
+            {
+                //Goto row was activated. As soon as a corresponding ID is found, change selection to it.
```
Comment (katalash) [inline StudioCore/ParamEditor/ParamEditorScreen.cs:1160]: Space after "//"

File: StudioCore/Scene/RenderableProxy.cs [hunk 14]
```diff
@@ -670,6 +672,7 @@ public void OnProviderAvailable()
                         }
                     }
                 }
+                MeshIndexCount = _meshProvider.IndexCount; //Used for model markers for old Navmeshes. Is this the best place for this check??
```
Comment (katalash) [inline StudioCore/Scene/RenderableProxy.cs:675]: This doesn't seem like the right place. Is this needed for navmeshes to render?

File: StudioCore/TextEditor/FMGBank.cs [hunk 15]
```diff
@@ -867,6 +867,7 @@ public static void SaveFMGs()
                 Utils.WriteWithBackup(AssetLocator.GameRootDirectory,
                     AssetLocator.GameModDirectory, menuMsgPathDest.AssetPath, (BND4)fmgBinderMenu);
             }
+            TaskManager.UpdateLastActionString("Saved FMGs");
```
Comment (katalash) [inline StudioCore/TextEditor/FMGBank.cs:870]: again task manager global state

File: StudioCore/TextEditor/TextEditorScreen.cs [hunk 16]
```diff
@@ -335,12 +335,12 @@ private void FMGSearchLogic()
...
                         else
-                            searchEntries = _cachedEntries;//FMGBank.GetItemFMGEntriesByType(_activeItemCategory, FMGBank.ItemType.Title).ToList()
+                            searchEntries = _cachedEntries;
 
                         foreach (var entry in searchEntries)
                         {
-                            if (entry.ID.ToString().Contains(_FMGsearchStr, StringComparison.CurrentCultureIgnoreCase))
-                                matches.Add(entry);
+                            if (entry.ID.ToString().Contains(_FMGsearchStr))
+                                matches.Add(entry); //ID matches
```
Comment (katalash) [inline StudioCore/TextEditor/TextEditorScreen.cs:343]: We're moving to case sensitive? Also added comment doesn't seem needed and I'd prefer comments to be on their own line almost always.

**Review by katalash — CHANGES_REQUESTED** [review]

File: StudioCore/MapStudioNew.cs (hunk 2 above)
Comment (kingborehaha) [reply StudioCore/MapStudioNew.cs:1053]: Previously, Ctrl+S hotkey only worked when the top left file menu was already open. This makes it so Ctrl+S works at all times (except for when viewport is being actively used, though the viewport check is pretty sloppy)

File: StudioCore/Scene/RenderableProxy.cs (hunk 14 above)
Comment (kingborehaha) [reply StudioCore/Scene/RenderableProxy.cs:675]: Yes, for DS1- navmesh rendering. Model marker system checks mesh's submesh count, which DS1 navmeshes don't use which causes rendering to collapse (and I failed to figure out why exactly rendering failed because of this, but I think disposing meshes was going wrong somewhere - a pre-me problem)

Checking index count seemed like a pretty stupid hot-glue solution (even to me) but in the interest of maintaining the (poorly implemented, but incredibly useful) model marker system while restoring navmesh rendering functionality.

So, a simple change to make to make things work in lieu of the inevitable model marker system rewrite that somebody smarter than me would have to do.

File: StudioCore/MsbEditor/PropertyEditor.cs (hunk 6 above)
Comment (kingborehaha) [reply StudioCore/MsbEditor/PropertyEditor.cs:394]: Does it really need to be cast in the first place though? It didn't seem like it did, maybe I'm overlooking something though. I left the stripped line + comment in there in case I was wrong.

File: StudioCore/Editor/TaskManager.cs (hunk 1 above)
Comment (kingborehaha) [reply StudioCore/Editor/TaskManager.cs:15]: Will strip out this system.

File: StudioCore/MapStudioNew.cs (hunk 3 above)
Comment (kingborehaha) [reply StudioCore/MapStudioNew.cs:684]: I don't think "Save active editor" will be clear enough for the user, at least based on confusion I've seen expressed by folk in the past. I don't know what future plans entail, but is this not acceptable at the current stage? Does more good than bad, I'd say.

File: StudioCore/ParamEditor/ParamEditorScreen.cs (hunk 9 above)
Comment (kingborehaha) [reply StudioCore/ParamEditor/ParamEditorScreen.cs:1068]: Gonna keep comment spacing in mind in the future, but for now how does a separate pull request that fixes the 1000+ current non-spaced comments sound?

File: StudioCore/TextEditor/TextEditorScreen.cs (hunk 16 above)
Comment (kingborehaha) [reply StudioCore/TextEditor/TextEditorScreen.cs:343]: This specifically searches IDs, so case check was unnecessary (though I suppose any impact is beyond negligible). ID match comment was to note that this was for ID searching (I'll move the comment)

File: StudioCore/MsbEditor/Entity.cs (hunk 4 above)
Comment (kingborehaha) [reply StudioCore/MsbEditor/Entity.cs:460]: I don't know enough to make a proper solution. This does serve as a functional solution (I was capable of doing) that lets the system work in a larger majority of situations, and it has behaved fast & fine in every common case I tested. 
Ref scanning/MSB search aren't able to go through structs as-is, so bigger changes need to happen in the future anyway and a rewrite at that time might be more appropriate.
